### PR TITLE
[WebGL] Support packed Conv2DBackpropInput

### PR DESCRIPTION
**Pull request by Linchenn**
With this PR:
Conv2DBackpropInput op is accelerated ~2x, partially fixed https://github.com/tensorflow/tfjs/issues/5197.

Model | Before(ms) | After(ms)
-- | -- | -- |
BlazePoseDetector | 23.1 | 17.5 
ArPortraitDepth | 52.3 | 23.9 (also set EXP_CONV to true)

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7339)

### CODE REVIEW

File: tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC. All Rights Reserved.
```
Comment (qjia7) [inline tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts:3]: nit: I remember @mattsoulanille ever mentioned `All Rights Reserved.` is not needed anymore for new generated file.

**Review by qjia7 — APPROVED** [review]
Clever algorithm and great work!

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7339)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn)

___
*[`tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts` line 55 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7339#-NNbiT3-BLjJgEaYDrEB:-NNbiT3-BLjJgEaYDrEC:b-4vhuky) ([raw file](https://github.com/tensorflow/tfjs/blob/062a3f5be3ae12f69dfcf0129b2e5349f6783fbe/tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts#L55)):*
> ```XML
> 
>         //intialize dotProd with a small epsilon seems to reduce GPU accuracy loss.
>         vec4 dotProd = vec4(0.000000000000001);
> ```

use an JS constant to interpolate in this value.
___
*[`tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts` line 90 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7339#-NNbj0pzB0a7jAf9N6dc:-NNbj0pzB0a7jAf9N6dd:b-pdywo) ([raw file](https://github.com/tensorflow/tfjs/blob/062a3f5be3ae12f69dfcf0129b2e5349f6783fbe/tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts#L90)):*
> ```XML
>               dotProd.xy += vec2(dot(dyValue, wValue.xy), dot(dyValue, wValue.zw)) * idyCVal;
> 
>               dySample = getDy(batch, idyR, idyC2, d2);
> ```

might be good to move all memory accesses together.
___
*[`tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts` line 90 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7339#-NNbjPce4xa1HqBDL_Oh:-NNbjPce4xa1HqBDL_Oi:b-s50bqt) ([raw file](https://github.com/tensorflow/tfjs/blob/062a3f5be3ae12f69dfcf0129b2e5349f6783fbe/tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts#L90)):*
> ```XML
>               dotProd.xy += vec2(dot(dyValue, wValue.xy), dot(dyValue, wValue.zw)) * idyCVal;
> 
>               dySample = getDy(batch, idyR, idyC2, d2);
> ```

can idyC2 could be the same as idyC ? only read again if idyC2 != idyC ?

File: tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts (hunk 1 above)
Comment (Linchenn) [reply tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts]: Nice catch, thanks!

**Review by Linchenn — COMMENTED** [review]
Thank you Ping. After tuning the shader, the inference time for the following op is improved from ~2.8928ms to ~2.6809903999999984ms on a Linux workstation:

```js
const x = tf.ones([1,8,6,256]);
const w = tf.ones([4,4,256,256]);

s = (await tf.profile(()=>tf.conv2dTranspose(x, w, [1,16,12,256], 1, 'valid'))).kernels[0].kernelTimeMs
```

After benchmarking, the improvement is mainly from add a condition for reading `dySample2`, from ~2.8928ms to ~2.711304960000001ms

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7339#-:-NNchQ0e5Omuvc-fihqi:b-3uh2z5)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @mattsoulanille and @pyu10055)

___
*[`tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts` line 55 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7339#-NNbiT3-BLjJgEaYDrEB:-NNcjjLwBoDdY-Ec9KQN:b577o6f) ([raw file](https://github.com/tensorflow/tfjs/blob/062a3f5be3ae12f69dfcf0129b2e5349f6783fbe/tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts#L55)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

use an JS constant to interpolate in this value.
</blockquote></details>

Removed it as we discussed, because this initial value currently is unnecessary, but we could keep this in mind when facing some correctness issues later.
___
*[`tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts` line 90 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7339#-NNbjPce4xa1HqBDL_Oh:-NNchM9O7oiyhpjNOEtB:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/062a3f5be3ae12f69dfcf0129b2e5349f6783fbe/tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts#L90)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

can idyC2 could be the same as idyC ? only read again if idyC2 != idyC ?
</blockquote></details>

Done.
___
*[`tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts` line 90 at r1](https://reviewable.io/reviews/tensorflow/tfjs/7339#-NNbj0pzB0a7jAf9N6dc:-NNchO0D5l1MTMvyeYdI:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/062a3f5be3ae12f69dfcf0129b2e5349f6783fbe/tfjs-backend-webgl/src/conv_backprop_packed_gpu.ts#L90)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

might be good to move all memory accesses together.
</blockquote></details>

Done.

**Review by pyu10055 — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7339)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @mattsoulanille)

**Pull request merged**